### ISSUE TITLE
Declare SMT variables lazily?

### ISSUE THREAD

**Issue by ruhler** [ruhler/smten#16]
Currently we declare every variable that could possibly appear in the query assertion to the SMT solver, even if that variable is not used. This causes strange looking debug outputs, for example, where hundreds of variables are declared, then the value 'True' is asserted, and all those variables are given arbitrary assignments.

What if we instead declared SMT variables lazily? That is, only declare SMT variables used in the assertion as we need them.

More specifically, the changes I foresee to implement this are:
- The 'solve' function no longer needs to take a list of variables. The list is implicit in the assertion.
- The assert traversal will allocate variables as needed.
- Defaulting of variables is done at the time of lookup in the model, not at the time of model formation.

Potential Benefits:
- Less time and space spent keeping track of what the free variables are.
- Less time spent declaring free variables.
- Cleaner debug outputs

Potential Downsides:
- We are putting yet more things into the assert traversal, though one could argue it already does declaration of variables, so this isn't any harder.
- We probably want some way to distinguish between user variables, which need to be part of the generated model, and internal variables, which are created internally, and don't need to be looked up or made part of the model.

In short, I feel like it could improve performance (though I doubt in any significant way), at no runtime cost, and with the code clarity degradation at least matched by its improvement elsewhere.

**Comment by ruhler** [comment]
I implemented this. It was easy, didn't involve any nasty code, seemed to clean things up, and appears to be a win for all aspects of performance (though I'm still not convinced the performance here is in any way significant).
